core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1153,7 +1153,16 @@ class DAGScheduler(
     }
 
     if (!stageIdToStage.contains(task.stageId)) {
-      // Skip all the actions if the stage has been cancelled.
+      logInfo("skip normal actions as stage cancelled")
+      // Need to handle tasks coming in late (speculative and jobs killed)
+      // post a task end event so accounting for things manually tracking tasks work.
+      // This really should be something other then success since the other speculative task
+      // finished first.
```

Comment:
by the way you should put SPARK-13054 in the title of this patch if you plan to do that here.